ledger/voters.go

```diff
@@ -173,25 +180,52 @@ func (vt *votersTracker) newBlock(hdr bookkeeping.BlockHeader) {
 		return
 	}
 
-	vt.removeOldVoters(hdr)
-
 	// This might be a block where we snapshot the online participants,
 	// to eventually construct a vector commitment in a later
 	// block.
-	r := uint64(hdr.Round)
-	if (r+proto.StateProofVotersLookback)%proto.StateProofInterval != 0 {
+	r := hdr.Round
+	if (uint64(r)+proto.StateProofVotersLookback)%proto.StateProofInterval != 0 {
 		return
 	}
 
-	_, ok := vt.votersForRoundCache[basics.Round(r)]
-	if ok {
+	_, exists := vt.getVoters(r)
+	if exists {
 		vt.l.trackerLog().Errorf("votersTracker.newBlock: round %d already present", r)
 	} else {
 		vt.loadTree(hdr)
 	}
 
 }
 
+func (vt *votersTracker) prepareCommit(dcc *deferredCommitContext) error {
+	vt.commitListenerMu.RLock()
+	defer vt.commitListenerMu.RUnlock()
+
+	if vt.commitListener == nil {
+		return nil
+	}
+
+	commitListener := vt.commitListener
+	for round := dcc.oldBase; round <= dcc.newBase; round++ {
+		err := commitListener.OnPrepareVoterCommit(round, vt)
+		if err != nil {
+			return err
```

Comment:
Should we really fail the entire ledger's commit if we have an error? I'm not sure. Your thoughts? @algoidan @algorandskiy

Reply:
Idan and I have decided that the answer is probably no, should still discuss this.